src/Microsoft.ML.Data/Transforms/KeyToVectorTransform.cs

```diff
@@ -206,393 +187,605 @@ public override void Save(ModelSaveContext ctx)
+            for (int i = 0; i < ColumnPairs.Length; i++)
+            {
+                ctx.Writer.Write(_sizes[i]);
+                ctx.Writer.Write(_valueCounts[i]);
+            }
         }
 
-        public override bool CanSavePfa => true;
-        public override bool CanSaveOnnx => true;
+        // Factory method for SignatureLoadModel.
+        public static KeyToVectorTransform Create(IHostEnvironment env, ModelLoadContext ctx)
+        {
+            Contracts.CheckValue(env, nameof(env));
+            var host = env.Register(RegistrationName);
+
+            host.CheckValue(ctx, nameof(ctx));
+            ctx.CheckAtModel(GetVersionInfo());
+
+            return new KeyToVectorTransform(host, ctx);
+        }
 
-        protected override JToken SaveAsPfaCore(BoundPfaContext ctx, int iinfo, ColInfo info, JToken srcToken)
+        private KeyToVectorTransform(IHost host, ModelLoadContext ctx)
+          : base(host, ctx)
         {
-            Contracts.AssertValue(ctx);
-            Contracts.Assert(0 <= iinfo && iinfo < Infos.Length);
-            Contracts.Assert(Infos[iinfo] == info);
-            Contracts.AssertValue(srcToken);
-            Contracts.Assert(CanSavePfa);
-
-            int keyCount = info.TypeSrc.ItemType.KeyCount;
-            Host.Assert(keyCount > 0);
-            // If the input type is scalar, we can just use the fanout function.
-            if (!info.TypeSrc.IsVector)
-                return PfaUtils.Call("cast.fanoutDouble", srcToken, 0, keyCount, false);
-
-            JToken arrType = PfaUtils.Type.Array(PfaUtils.Type.Double);
-            if (_concat[iinfo])
+            var columnsLength = ColumnPairs.Length;
+            // *** Binary format ***
+            // <base>
+            // for each added column
+            //   byte: bag as 0/1
+            // for each added column
+            //  int: keyCount
+            //  int: valueCount
+            _bags = new bool[columnsLength];
+            _sizes = new int[columnsLength];
+            _valueCounts = new int[columnsLength];
+
+            _bags = ctx.Reader.ReadBoolArray(columnsLength);
+            for (int i = 0; i < columnsLength; i++)
             {
-                // The concatenation case. We can still use fanout, but we just append them all together.
-                return PfaUtils.Call("a.flatMap", srcToken,
-                    PfaContext.CreateFuncBlock(new JArray() { PfaUtils.Param("k", PfaUtils.Type.Int) },
-                    arrType, PfaUtils.Call("cast.fanoutDouble", "k", 0, keyCount, false)));
+                _sizes[i] = ctx.Reader.ReadInt32();
+                _valueCounts[i] = ctx.Reader.ReadInt32();
             }
+        }
+
+        public static IDataTransform Create(IHostEnvironment env, IDataView input, params ColumnInfo[] columns) =>
+             new KeyToVectorTransform(env, input, columns).MakeDataTransform(input);
+
+        // Factory method for SignatureDataTransform.
```

Comment:
>// Fa [](start = 8, length = 5)

same here.

Reply:
I'll rather make component catalog to support internal Create methods

---
In reply to: [216121776](https://github.com/dotnet/machinelearning/pull/858#discussion_r216121776) [](ancestors = 216121776)